File: Java/프로그래머스/2/148653. 마법의 엘리베이터/마법의 엘리베이터.java
class Solution {
    public int solution(int storey) {
        if (storey < 10) {
            return min(storey, 10 - storey + 1);
        }
        
        int div = storey / 10, remainder = storey % 10;
        return min(solution(div) + remainder, solution(div + 1) + 10 - remainder);
    }
    
    private int min(int A, int B) {
        if (A > B) return B;
        return A;
    }
}

File: Python3/프로그래머스/2/148653. 마법의 엘리베이터/마법의 엘리베이터.py
def go_up(floor):
    diff = 10 - (floor % 10)
    return ((floor + diff) // 10 , diff)


def go_down(floor):
    diff = floor % 10
    return ((floor - diff) // 10, diff)


def solution(storey):
    answer = 100
    stack = [(storey, 0)]
    while stack:
        floor, cnt = stack.pop()
        
        while floor % 10 == 0:
            floor //= 10
            
        if floor > 10:            
            n_floor, diff = go_down(floor)
            stack.append((n_floor, cnt + diff))
            n_floor, diff = go_up(floor)
            stack.append((n_floor, cnt + diff))
        else:
            answer = min(answer, cnt + min(floor, 10 - floor + 1))
        
    return answer